core/src/main/java/com/linecorp/armeria/internal/common/DefaultSplitHttpResponse.java

```diff
@@ -69,202 +44,33 @@
         EMPTY_TRAILERS.doComplete(HttpHeaders.of());
     }
 
-    private final HeadersFuture<ResponseHeaders> headersFuture = new HeadersFuture<>();
-    private final BodySubscriber bodySubscriber = new BodySubscriber();
-    private final HttpResponse response;
-    private final EventExecutor upstreamExecutor;
-
-    @Nullable
-    private volatile HeadersFuture<HttpHeaders> trailersFuture;
-    private volatile boolean wroteAny;
+    private final SplitHttpResponseBodySubscriber bodySubscriber;
 
     public DefaultSplitHttpResponse(HttpResponse response, EventExecutor executor) {
-        this.response = requireNonNull(response, "response");
-        upstreamExecutor = requireNonNull(executor, "executor");
-
-        response.subscribe(bodySubscriber, upstreamExecutor, SubscriptionOption.values());
-    }
-
-    @Override
-    public final CompletableFuture<ResponseHeaders> headers() {
-        return headersFuture;
-    }
-
-    @Override
-    public final StreamMessage<HttpData> body() {
-        return this;
-    }
-
-    @Override
-    public final CompletableFuture<HttpHeaders> trailers() {
-        HeadersFuture<HttpHeaders> trailersFuture = this.trailersFuture;
-        if (trailersFuture != null) {
-            return trailersFuture;
-        }
-
-        trailersFuture = new HeadersFuture<>();
-        if (trailersFutureUpdater.compareAndSet(this, null, trailersFuture)) {
-            return trailersFuture;
-        } else {
-            return this.trailersFuture;
-        }
-    }
-
-    @Override
-    public boolean isOpen() {
-        return response.isOpen();
-    }
-
-    @Override
-    public boolean isEmpty() {
-        return !isOpen() && !wroteAny;
-    }
-
-    @Override
-    public long demand() {
-        return response.demand();
-    }
-
-    @Override
-    public CompletableFuture<Void> whenComplete() {
-        return response.whenComplete();
+        this(response, executor, new SplitHttpResponseBodySubscriber(response, executor));
     }
 
-    @Override
-    public EventExecutor defaultSubscriberExecutor() {
-        return upstreamExecutor;
+    private DefaultSplitHttpResponse(HttpResponse response, EventExecutor executor,
+                                     SplitHttpResponseBodySubscriber bodySubscriber) {
+        super(response, executor, bodySubscriber);
+        this.bodySubscriber = requireNonNull(bodySubscriber, "bodySubscriber");
```

Comment:
We usually put `requireNonNull` at the first line of the method.
But, I think we can remove null check completely because this is a private method. 😉 
```suggestion
        this.bodySubscriber = bodySubscriber;
```